core/ArchiveProcessor/Loader.php

```diff
@@ -388,6 +389,43 @@ public function canSkipThisArchive()
         $isArchivingForcedWhenNoVisits = $this->shouldArchiveForSiteEvenWhenNoVisits();
         $hasSiteVisitsBetweenTimeframe = $this->hasSiteVisitsBetweenTimeframe($idSite, $params->getPeriod());
         $hasChildArchivesInPeriod = $this->dataAccessModel->hasChildArchivesInPeriod($idSite, $params->getPeriod());
+        $hasSegment = !$params->getSegment()->isEmpty();
+
+        if ($hasSegment) {
+            $periodEnd = $params->getPeriod()->getDateEnd();
+            $segmentHash = $params->getSegment()->getHash();
+            /** @var SegmentArchiving */
+            $segmentArchiving = StaticContainer::get(SegmentArchiving::class);
+            $segmentInfo = $segmentArchiving->findSegmentForHash($segmentHash, $idSite);
+
+            if ($segmentInfo) {
+                $segmentArchiveStartDate = $segmentArchiving->getReArchiveSegmentStartDate($segmentInfo);
+
+                if ($segmentArchiveStartDate !==null && $segmentArchiveStartDate->isLater($periodEnd->getEndOfDay())) {
+                    $doneFlag = Rules::getDoneStringFlagFor(
+                        [$idSite],
+                        $params->getSegment(),
+                        $params->getPeriod()->getLabel(),
+                        $params->getRequestedPlugin()
+                    );
+
+                    $table = Common::prefixTable('archive_invalidations');
+                    $sql = "SELECT idinvalidation FROM `$table` WHERE idsite = ? AND date1 = ? AND date2 = ? AND `period` = ? AND `name` = ?";
+
+                    $idInvalidation = Db::fetchOne($sql, [
+                        $idSite,
+                        $params->getPeriod()->getDateStart()->toString(),
+                        $params->getPeriod()->getDateEnd()->toString(),
+                        $params->getPeriod()->getId(),
+                        $doneFlag
+                    ]);
+
+                    if (empty($idInvalidation)) {
+                        return true;
+                    }
+                }
+            }
+        }
```

Comment:
Can we put all this logic into a descriptively named function, like the previous function calls are? Would it also be possible to write integration tests for the new method?

Reply:
I refactored and added tests 👍